rtCommon/bidsIncremental.py

```diff
@@ -0,0 +1,588 @@
+                 datasetMetadata: dict = None):
+        """
+        Initializes a BIDS Incremental object with provided image and metadata.
+
+        Args:
+            image: NIfTI image as an NiBabel NiftiImage or PyBids BIDSImageFile
+            imageMetadata: Metadata for image, which must include all variables
+                in BidsIncremental.REQUIRED_IMAGE_METADATA.
+            datasetMetadata: Top-level dataset metadata for the BIDS dataset
+                to be placed in a dataset_description.json. Defaults to None and
+                a default description is used.
+
+        Raises:
+            MissingMetadataError: If any required metadata is missing.
+            TypeError: If the image is not an Nibabel Nifti1Image or
+                Nifti2Image.
+
+        Examples:
+            >>> import nibabel as nib
+            >>> imageMetadata = {'subject': '01', 'task': 'test',
+                                 'suffix': 'bold', 'datatype': 'func',
+                                 'RepetitionTime': 1.5, 'EchoTime': 1}
+            >>> image = nib.load('/tmp/testfile.nii')
+            >>> datasetMetadata = {'Name': 'Example Dataset',
+                                   'BIDSVersion': '1.5.1',
+                                   'Authors': 'The RT-Cloud Authors'}
+            >>> incremental = BidsIncremental(image, imageMetadata,
+                datasetMetadata)
+            >>> print(incremental)
+            "Image shape: (64, 64, 27, 1); Metadata Key Count: 6; BIDS-I
+            Version: 1"
+        """
+        """ Do basic input validation """
+        # IMAGE
+        validTypes = [nib.Nifti1Image, nib.Nifti2Image, BIDSImageFile]
+        if image is None or type(image) not in validTypes:
+            raise TypeError("Image must be Nibabel Nifti1 or Nifti22 (got "
+                            f"{type(image)}")
+        if type(image) is BIDSImageFile:
+            image = image.get_image()
+
+        # DATASET METADATA
+        if datasetMetadata is not None:
+            missingFields = [field for field in DATASET_DESC_REQ_FIELDS
+                             if datasetMetadata.get(field, None) is None]
+            if missingFields:
+                raise MissingMetadataError(
+                    f"Dataset description needs: {str(missingFields)}")
+
+        """ Process, validate, and store image metadata """
+        imageMetadata = self._preprocessMetadata(imageMetadata)
+        self._exceptIfMissingMetadata(imageMetadata)
+        self._imgMetadata = self._postprocessMetadata(imageMetadata)
+
+        """ Store dataset metadata """
+        if datasetMetadata is None:
+            self.datasetMetadata = DEFAULT_DATASET_DESC
+        else:
+            self.datasetMetadata = deepcopy(datasetMetadata)
+
+        """ Validate and store image """
+        # Remove singleton dimensions
+        self.image = nib.funcs.squeeze_image(image)
+
+        imageShape = self.imageDimensions
```

Comment:
GW comment: are you missing () here, i.e. self.imageDimensions()? A secondary comment, I think it's not common practice to call a class method from within __init__, because it could be something required by the method hasn't been init'd yet. Perhaps retrieve the dimensions directly?

Reply:
TODO: Add comment about needing to add image as class member first